### DISCUSSION TITLE
Should be able to add stylesheets dynamically (app router)

### DISCUSSION THREAD

**Discussion by maxswjeon** [vercel/next.js#49744] [Ideas]
### Goals

Should be able to add stylesheets with current `Metadata` or `generateMetadata` structure

### Background

There are several use cases that needs to dynamically add stylesheets, such as user-created theme, lots of font loading etc. Currently, there is a limited way to do this, by dynamically adding `<link>` tags in the root layout (`apps/layout.ts`).

Some of the use cases only requires to add `<link>` tags only on specific routes only. However, since the `<head>` tag can be appear on the root layout only, so we are only able to add `<link>` to all the routes.

Take [this example](https://github.com/maxswjeon/editorial). This project is a CMS that allows user to upload a theme, and apply it to the site. User should be able to add custom fonts or custom CSS theme. All of those cannot be known at build time. The route handler in the `app/theme/route.ts` handles file I/O.  

However, since the loading of the custom fonts and the custom CSS theme is occuring on the root layout, the custom fonts and custom CSS is injected to all routes, which I don't want to do. (I don't want to allow custom fonts and custom CSS on admin routes). 

### Proposal

There are several ways to achieve this.

1. Add a new metadata prop, `stylesheet`, with the type below.
   ```typescript
   stylesheet?: {
     crossOrigin?: "anonymous" | "use-credentials";
     href: string;
   }[];
   ```
2. Change the `other` metadata prop to allow add not only `<meta>` tags.
   ```typescript
   other?: {
     type: "link" | "meta";
     [key: string]: string;
   }[]
   ```
3. Reintroduce `<Head>` component.

I'm willing to contribute, but unable to contribute before the discussion starts

**Comment by Filippinifra** [comment]
update on this?

**Reply by elylucas** [reply]
+1, this is a regression coming from pages router

**Comment by olayway** [comment]
I'm also trying to do this and haven't found a way

**Comment by rnnyrk** [comment]
Trying to find a way to add dynamic stylesheets based on a theme as well. No luck so far, except just rendering the stylesheet inside the body with a Link. Unwanted behaviour, but does the trick.

**Comment by candidodmv** [comment]
I'm looking for the same thing. Please @leerob, could you check out this feature and provide some workaround to solve this limitation?

**Reply by leerob** [reply]
With the latest React features, you can just include the links in the component.

- https://react.dev/blog/2024/04/25/react-19#support-for-stylesheets
- https://www.youtube.com/watch?v=IBBN-s77YSI


These are all available in the Next.js App Router today.

**Reply by candidodmv** [reply]
I'm gonna check out right now. Thank you @leerob.
